stellargraph/connector/neo4j/sampler.py

```diff
@@ -0,0 +1,202 @@
+]
+
+
+import numpy as np
+import warnings
+from collections import defaultdict, deque
+
+from ...core.schema import GraphSchema
+from ...core.graph import StellarGraph
+from ...data.explorer import GraphWalk
+from ...core.experimental import experimental
+
+
+@experimental(reason="the class is not fully tested")
+class Neo4JSampledBreadthFirstWalk(GraphWalk):
+    """
+    Breadth First Walk that generates a sampled number of paths from a starting node.
+    It can be used to extract a random sub-graph starting from a set of initial nodes from Neo4J database.
+    """
+
+    def run(self, neo4j_graphdb, nodes=None, n=1, n_size=None, seed=None):
+        """
+        Send queries to Neo4J graph databases and collect sampled breadth-first walks starting from the root nodes.
+
+        Args:
+            neo4j_graphdb: (py2neo.Graph) the Neo4J Graph Database object
+            nodes (list): A list of root node ids such that from each node n BFWs will be generated up to the
+            given depth d.
+            n_size (int): The number of neighbouring nodes to expand at each depth of the walk. Sampling of
+            n (int, default 1): Number of walks per node id.
+            neighbours with replacement is always used regardless of the node degree and number of neighbours
+            requested.
+            seed (int, optional): Random number generator seed; default is None
+
+        Returns:
+            A list of lists, each list is a sequence of sampled node ids at a certain hop.
+        """
+
+        def bfs_neighbor_query(sampling_direction):
+            direction = {"BOTH": "--", "IN": "<--", "OUT": "-->"}
+
+            return f"""
+                // expand the list of node id in seperate rows of ids.
+                UNWIND {{node_id_list}} AS node_id
+
+                // for each node id in every row, collect the random list of its neighbors.
+                CALL apoc.cypher.run(
+
+                    'MATCH(cur_node) WHERE id(cur_node) = {{node_id}}
+
+                    // find the neighbors
+                    MATCH (cur_node){direction[sampling_direction]}(neighbors)
+
+                    // put all ids into a list
+                    WITH CASE collect(id(neighbors)) WHEN [] THEN [null] ELSE collect(id(neighbors)) END AS in_neighbors_list
+
+                    // pick random nodes with replacement
+                    WITH apoc.coll.randomItems(in_neighbors_list, {{num_samples}}, True) AS in_samples_list
+
+                    RETURN in_samples_list',
+                    {{ node_id: node_id, num_samples: {{num_samples}}  }}) YIELD value
+
+                RETURN apoc.coll.flatten(collect(value.in_samples_list)) as next_samples
+            """
+
+        samples = [[head_node for head_node in nodes for _ in range(n)]]
+        neighbor_query = bfs_neighbor_query(sampling_direction="BOTH")
+
+        for num_sample in n_size:
+            cur_nodes = samples[-1]
+            result = neo4j_graphdb.run(
```

Comment:
It looks like this code has gone from doing one round-trip to neo4j by sending the query as one huge string, to doing `len(n_size)` trips, each one sending back the list of nodes. Don't change it back just yet, but could you explain why you changed it?

Reply:
Yes that's absolutely right! 

The main motivation is to make the queries neater and reusable across the module. Joining queries together is possible but this leads to much more syntax and is bad when maintaining the code. 

It is obviously easier to let Python stack the resulting hops together, but yes, this does come with the tradeoff of having to execute more transactions between Python and the databases. This ended up sending O(n) and O(2^n) queries to the graph databases (n is the number of hops). In practice, we experienced around 10% slower than sending one query that does all, which I think is acceptable.

Reply:
> In practice, we experienced around 10% slower than sending one query that does all

"We"? Did multiple people test it?

Also, were you testing with a local neo4j, or over a network connection? If it's local, the overhead of connecting to it will be very small, whereas if it's a neo4j running on a different computer, there's likely to be noticeably more overhead.

In any case, it's fine to leave this as is, but it would be good to have a comment, like:

```python
# this does O(number of hops) queries to the database, because the code is cleanest like that
```

(With `O(number of hops)` -> `O(2^number of hops)` for the other one)

> Because n does not exceed 3 for most of the times, I guess it is still acceptable!

Doing 8 queries (with significant amounts of data set back and forth) instead of 1 is a big expansion. But yes, it's probably acceptable for now.